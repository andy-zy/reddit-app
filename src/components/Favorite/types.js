// @flow

export type PropsT = {
  isFavorite: boolean,
  toggleFavorite: Function,
}