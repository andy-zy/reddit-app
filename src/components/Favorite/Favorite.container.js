import { connect } from 'react-redux'

import Favorite from './Favorite'
import { activeArticleSelector, favoritesByIdSelector } from '../../domain/selectors'
import { addFavorite, removeFavorite } from '../../domain/actions'

const mapStateToProps = (state) => ({
  activeArticle: activeArticleSelector(state),
  favoritesById: favoritesByIdSelector(state),
})

const mapActionsToProps = {
  addFavorite,
  removeFavorite,
}

const mergeProps = (stateProps, dispatchProps, ownProps) => {
  const article = ownProps.article || stateProps.activeArticle
  const isFavorite = !!article && stateProps.favoritesById.includes(article.id)
  const toggleFavorite = () => isFavorite ?
    dispatchProps.removeFavorite(article.id) : dispatchProps.addFavorite(article)

  return {
    isFavorite,
    toggleFavorite,
  }
}

export default connect(
  mapStateToProps,
  mapActionsToProps,
  mergeProps,
)(Favorite)