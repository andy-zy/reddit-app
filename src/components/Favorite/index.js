import Favorite from './Favorite.container'

export default Favorite