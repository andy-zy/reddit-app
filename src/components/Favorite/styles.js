import {
  StyleSheet,
} from 'react-native'

export default StyleSheet.create({
  container: {
    marginRight: 10,
  },
  star: {
    width: 24,
    height: 24,
  },
})