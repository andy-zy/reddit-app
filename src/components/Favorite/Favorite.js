// @flow
import React from 'react'
import {
  Image,
  TouchableHighlight,
} from 'react-native'

import type { PropsT } from './types'

import styles from './styles'

const Favorite = ({ isFavorite, toggleFavorite }: PropsT) => (
  <TouchableHighlight onPress={toggleFavorite} style={styles.container}>
    {
      isFavorite ? (
        <Image source={require('../../images/marked_favorite.png')} style={styles.star} />
      ) : (
        <Image source={require('../../images/unmarked_favorite.png')} style={styles.star} />
      )
    }
  </TouchableHighlight>
)

export default Favorite