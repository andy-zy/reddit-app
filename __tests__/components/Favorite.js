import 'react-native'
import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import Favorite from '../../src/components/Favorite/Favorite'

describe('Favorite', () => {
  let props

  beforeEach(() => {
    props = {
      isFavorite: false,
      toggleFavorite: jest.fn(),
    }
  })

  it('renders correctly non-favorite layout', () => {
    const tree = renderer.create(<Favorite {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly favorite layout', () => {
    const tree = renderer.create(<Favorite {...props} isFavorite />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('onPress should call toggleFavorite', () => {
    const component = shallow(<Favorite {...props} />)
    component.simulate('press')
    expect(props.toggleFavorite).toHaveBeenCalled()
  })
})