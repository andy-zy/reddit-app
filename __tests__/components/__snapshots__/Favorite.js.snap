// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorite renders correctly favorite layout 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "marginRight": 10,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <Image
    source={
      Object {
        "testUri": "../../../src/images/marked_favorite.png",
      }
    }
    style={
      Object {
        "height": 24,
        "width": 24,
      }
    }
  />
</View>
`;

exports[`Favorite renders correctly non-favorite layout 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "marginRight": 10,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <Image
    source={
      Object {
        "testUri": "../../../src/images/unmarked_favorite.png",
      }
    }
    style={
      Object {
        "height": 24,
        "width": 24,
      }
    }
  />
</View>
`;
